File: android/src/main/java/com/audiocontext/nodes/filter/BiquadFilterNode.kt
package com.audiocontext.nodes.filter

import android.os.Build
import androidx.annotation.RequiresApi
import com.audiocontext.context.BaseAudioContext
import com.audiocontext.nodes.AudioNode
import com.audiocontext.parameters.AudioParam
import com.audiocontext.parameters.PlaybackParameters
import com.audiocontext.utils.Constants
import kotlin.math.*

//https://webaudio.github.io/Audio-EQ-Cookbook/audio-eq-cookbook.html
//https://github.com/LabSound/LabSound/blob/main/src/internal/src/Biquad.cpp

class BiquadFilterNode(context: BaseAudioContext): AudioNode(context) {
  override val numberOfInputs: Int = 1
  override val numberOfOutputs: Int = 1
  private val frequency: AudioParam = AudioParam(context, 350.0, Constants.MAX_FILTER_FREQUENCY, Constants.MIN_FILTER_FREQUENCY)
    get() = field
  private val detune: AudioParam = AudioParam(context, 0.0, Constants.MAX_DETUNE, -Constants.MAX_DETUNE)
    get() = field
  private val Q: AudioParam = AudioParam(context, 1.0, Constants.MAX_FILTER_Q, -Constants.MAX_FILTER_Q)
    get() = field
  private val gain: AudioParam = AudioParam(context, 0.0, Constants.MAX_FILTER_GAIN, Constants.MIN_FILTER_GAIN)
    get() = field
  private var filterType: FilterType = FilterType.LOWPASS

  // delayed samples
  private var y1 = 0.0
  private var y2 = 0.0
  private var x1 = 0.0
  private var x2 = 0.0
  // coefficients
  private var a1 = 0.0
  private var a2 = 0.0
  private var b0 = 1.0
  private var b1 = 0.0
  private var b2 = 0.0

  fun getFilterType(): String {
    return FilterType.toString(filterType)
  }

  fun setFilterType(type: String) {
    filterType = FilterType.fromString(type)
  }

  private fun reset() {
    y1 = 0.0
    y2 = 0.0
    x1 = 0.0
    x2 = 0.0
  }

  private fun setNormalizedCoefficients(a0: Double, a1: Double, a2: Double, b0: Double, b1: Double, b2: Double) {
    this.a1 = a1 / a0
    this.a2 = a2 / a0
    this.b0 = b0 / a0
    this.b1 = b1 / a0
    this.b2 = b2 / a0
  }

  private fun setLowpassCoefficients(cutOffFrequency: Double, Q: Double) {
    val cutOff = max(0.0, min(cutOffFrequency, Constants.NYQUIST_FREQUENCY))

    if (cutOff == 1.0) {
      setNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
    } else if (cutOff > 0.0) {
      val resonance = max(0.0, Q)
      val g = 10.0.pow(resonance / 20.0)
      val d = sqrt((4.0 - sqrt(16.0 - 16.0 / (g * g))) / 2.0)

      val theta = 2.0 * PI * cutOff
      val sn = 0.5 * d * sin(theta)
      val beta = 0.5 * (1.0 - sn) / (1.0 + sn)
      val gamma = (0.5 + beta) * cos(theta)
      val alpha = 0.25 * (0.5 + beta - gamma)

      val b0 = 2.0 * alpha
      val b1 = 4.0 * alpha
      val b2 = 2.0 * alpha
      val a1 = -2.0 * gamma
      val a2 = 2.0 * beta

      setNormalizedCoefficients(1.0, a1, a2, b0, b1, b2)
    } else {
      setNormalizedCoefficients(1.0, 0.0, 0.0, 0.0, 0.0, 0.0)
    }
  }

  private fun setHighpassCoefficients(frequency: Double, Q: Double) {
    val cutOff = max(0.0, min(frequency, 1.0))

    if (cutOff == 1.0) {
      setNormalizedCoefficients(1.0, 0.0, 0.0, 0.0, 0.0, 0.0)
    } else if (cutOff > 0.0) {
      val resonance = max(0.0, Q)
      val g = 10.0.pow(resonance / 20.0)
      val d = sqrt((4.0 - sqrt(16.0 - 16.0 / (g * g))) / 2.0)

      val theta = 2.0 * PI * cutOff
      val sn = 0.5 * d * sin(theta)
      val beta = 0.5 * (1.0 - sn) / (1.0 + sn)
      val gamma = (0.5 + beta) * cos(theta)
      val alpha = 0.25 * (0.5 + beta - gamma)

      val b0 = 2.0 * alpha
      val b1 = -4.0 * alpha
      val b2 = 2.0 * alpha
      val a1 = -2.0 * gamma
      val a2 = 2.0 * beta

      setNormalizedCoefficients(1.0, a1, a2, b0, b1, b2)
    } else {
      setNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
    }
  }

  private fun setBandpassCoefficients(frequency: Double, Q: Double) {
    val cutOff = max(0.0, frequency)
    val resonance = max(0.0, Q)

    if (cutOff > 0 && cutOff < 1) {
      if (resonance > 0) {
        val w0 = PI * cutOff
        val alpha = sin(w0) / (2.0 * resonance)

        val b0 = alpha
        val b1 = 0.0
        val b2 = -alpha
        val a0 = 1.0 + alpha
        val a1 = -2.0 * cos(w0)
        val a2 = 1.0 - alpha

        setNormalizedCoefficients(a0, a1, a2, b0, b1, b2)
      } else {
        setNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
      }
    } else {
      setNormalizedCoefficients(1.0, 0.0, 0.0, 0.0, 0.0, 0.0)
    }
  }

  private fun setLowshelfCoefficients(frequency: Double, gain: Double) {
    val cutOff = max(0.0, min(frequency, 1.0))
    val A = 10.0.pow(gain / 40.0)

    if (cutOff == 1.0) {
      setNormalizedCoefficients(1.0, 0.0, 0.0, A * A, 0.0, 0.0)
    } else if (cutOff > 0.0) {
      val w0 = PI * cutOff
      val S = 1.0
      val alpha = sin(w0) / 2.0 * sqrt((A + 1.0 / A) * (1.0 / S - 1.0) + 2.0)

      val b0 = A * ((A + 1.0) - (A - 1.0) * cos(w0) + 2.0 * sqrt(A) * alpha)
      val b1 = 2.0 * A * ((A - 1.0) - (A + 1.0) * cos(w0))
      val b2 = A * ((A + 1.0) - (A - 1.0) * cos(w0) - 2.0 * sqrt(A) * alpha)
      val a0 = (A + 1.0) + (A - 1.0) * cos(w0) + 2.0 * sqrt(A) * alpha
      val a1 = -2.0 * ((A - 1.0) + (A + 1.0) * cos(w0))
      val a2 = (A + 1.0) + (A - 1.0) * cos(w0) - 2.0 * sqrt(A) * alpha

      setNormalizedCoefficients(a0, a1, a2, b0, b1, b2)
    } else {
      setNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
    }
  }

  private fun setHighshelfCoefficients(frequency: Double, gain: Double) {
    val cutOff = max(0.0, min(frequency, 1.0))
    val A = 10.0.pow(gain / 40.0)

    if (cutOff == 1.0) {
      setNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
    } else if (cutOff > 0.0) {
      val w0 = PI * cutOff
      val S = 1.0
      val alpha = sin(w0) / 2.0 * sqrt((A + 1.0 / A) * (1.0 / S - 1.0) + 2.0)

      val b0 = A * ((A + 1.0) - (A - 1.0) * cos(w0) + 2.0 * sqrt(A) * alpha)
      val b1 = -2.0 * A * ((A - 1.0) + (A + 1.0) * cos(w0))
      val b2 = A * ((A + 1.0) + (A - 1.0) * cos(w0) - 2.0 * sqrt(A) * alpha)
      val a0 = (A + 1.0) - (A - 1.0) * cos(w0) + 2.0 * sqrt(A) * alpha
      val a1 = 2.0 * ((A - 1.0) + (A + 1.0) * cos(w0))
      val a2 = (A + 1.0) - (A - 1.0) * cos(w0) - 2.0 * sqrt(A) * alpha

      setNormalizedCoefficients(a0, a1, a2, b0, b1, b2)
    } else {
      setNormalizedCoefficients(1.0, 0.0, 0.0, A*A, 0.0, 0.0)
    }
  }

  private fun setPeakingCoefficients(frequency: Double, Q: Double, gain: Double) {
    val cutOff = max(0.0, min(frequency, 1.0))
    val resonance = max(0.0, Q)
    val A = 10.0.pow(gain / 40.0)

    if (cutOff > 0 && cutOff < 1) {
      if (resonance > 0) {
        val w0 = PI * cutOff
        val alpha = sin(w0) / (2.0 * resonance)

        val b0 = 1.0 + alpha * A
        val b1 = -2.0 * cos(w0)
        val b2 = 1.0 - alpha * A
        val a0 = 1.0 + alpha / A
        val a1 = -2.0 * cos(w0)
        val a2 = 1.0 - alpha / A

        setNormalizedCoefficients(a0, a1, a2, b0, b1, b2)
      } else {
        setNormalizedCoefficients(1.0, 0.0, 0.0, A*A, 0.0, 0.0)
      }
    } else {
      setNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
    }
  }

  private fun setNotchCoefficients(frequency: Double, Q: Double) {
    val cutOff = max(0.0, min(frequency, 1.0))
    val resonance = max(0.0, Q)

    if (cutOff > 0 && cutOff < 1) {
      if (resonance > 0) {
        val w0 = PI * cutOff
        val alpha = sin(w0) / (2.0 * resonance)

        val b0 = 1.0
        val b1 = -2.0 * cos(w0)
        val b2 = 1.0
        val a0 = 1.0 + alpha
        val a1 = -2.0 * cos(w0)
        val a2 = 1.0 - alpha

        setNormalizedCoefficients(a0, a1, a2, b0, b1, b2)
      } else {
        setNormalizedCoefficients(1.0, 0.0, 0.0, 0.0, 0.0, 0.0)
      }
    } else {
      setNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
    }
  }

  private fun setAllpassCoefficients(frequency: Double, Q: Double) {
    val cutOff = max(0.0, min(frequency, 1.0))
    val resonance = max(0.0, Q)

    if (cutOff > 0 && cutOff < 1) {
      if (resonance > 0) {
        val w0 = PI * cutOff
        val alpha = sin(w0) / (2.0 * resonance)

        val b0 = 1.0 - alpha
        val b1 = -2.0 * cos(w0)
        val b2 = 1.0 + alpha
        val a0 = 1.0 + alpha
        val a1 = -2.0 * cos(w0)
        val a2 = 1.0 - alpha

        setNormalizedCoefficients(a0, a1, a2, b0, b1, b2)
      } else {
        setNormalizedCoefficients(1.0, 0.0, 0.0, -1.0, 0.0, 0.0)
      }
    } else {
      setNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
    }
  }

  @RequiresApi(Build.VERSION_CODES.N)
  private fun applyFilter() {
    var normalizedFrequency = frequency.getValueAtTime(context.getCurrentTime()) / Constants.NYQUIST_FREQUENCY

    if (detune.getValueAtTime(context.getCurrentTime()) != 0.0) {
      normalizedFrequency *= 2.0.pow(detune.getValueAtTime(context.getCurrentTime()) / 1200.0)
    }

    when (filterType) {
      FilterType.LOWPASS -> {
        setLowpassCoefficients(normalizedFrequency, Q.getValueAtTime(context.getCurrentTime()))
      }
      FilterType.HIGHPASS -> {
        setHighpassCoefficients(normalizedFrequency, Q.getValueAtTime(context.getCurrentTime()))
      }
      FilterType.BANDPASS -> {
        setBandpassCoefficients(normalizedFrequency, Q.getValueAtTime(context.getCurrentTime()))
      }
      FilterType.LOWSHELF -> {
        setLowshelfCoefficients(normalizedFrequency, gain.getValueAtTime(context.getCurrentTime()))
      }
      FilterType.HIGHSHELF -> {
        setHighshelfCoefficients(normalizedFrequency, gain.getValueAtTime(context.getCurrentTime()))
      }
      FilterType.PEAKING -> {
        setPeakingCoefficients(normalizedFrequency, Q.getValueAtTime(context.getCurrentTime()), gain.getValueAtTime(context.getCurrentTime()))
      }
      FilterType.NOTCH -> {
        setNotchCoefficients(normalizedFrequency, Q.getValueAtTime(context.getCurrentTime()))
      }
      FilterType.ALLPASS -> {
        setAllpassCoefficients(normalizedFrequency, Q.getValueAtTime(context.getCurrentTime()))
      }
    }
  }

  @RequiresApi(Build.VERSION_CODES.N)
  override fun process(playbackParameters: PlaybackParameters) {
    reset()
    applyFilter()
    for (i in playbackParameters.buffer.indices) {
      val input = playbackParameters.buffer[i]
      val output = b0 * input + b1 * x1 + b2 * x2 - a1 * y1 - a2 * y2

      x2 = x1
      x1 = input.toDouble()
      y2 = y1
      y1 = output

      playbackParameters.buffer[i] = output.toInt().toShort()
    }

    super.process(playbackParameters)
  }
}

File: packages/react-native-audio-api/common/cpp/audioapi/core/effects/BiquadFilterNode.cpp
/*
 * Copyright (C) 2010 Google Inc. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1.  Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 * 2.  Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 * 3.  Neither the name of Apple Computer, Inc. ("Apple") nor the names of
 *     its contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY APPLE AND ITS CONTRIBUTORS "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL APPLE OR ITS CONTRIBUTORS BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <audioapi/core/BaseAudioContext.h>
#include <audioapi/core/effects/BiquadFilterNode.h>
#include <audioapi/core/utils/Constants.h>
#include <audioapi/types/NodeOptions.h>

#include <algorithm>
#include <cmath>
#include <memory>
#include <numbers>

// https://webaudio.github.io/Audio-EQ-Cookbook/audio-eq-cookbook.html - math
// formulas for filters
// NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers, readability-magic-numbers, readability-identifier-length)

namespace audioapi {

namespace {

constexpr double kPi = std::numbers::pi_v<double>;
constexpr double kSqrt2 = std::numbers::sqrt2_v<double>;

} // namespace

BiquadFilterNode::BiquadFilterNode(
    const std::shared_ptr<BaseAudioContext> &context,
    const BiquadFilterOptions &options)
    : AudioNode(context, options),
      frequencyParam_(
          std::make_shared<AudioParam>(options.frequency, 0.0f, getNyquistFrequency(), context)),
      detuneParam_(
          std::make_shared<AudioParam>(
              options.detune,
              -OCTAVE_RANGE * LOG2_MOST_POSITIVE_SINGLE_FLOAT,
              OCTAVE_RANGE * LOG2_MOST_POSITIVE_SINGLE_FLOAT,
              context)),
      QParam_(
          std::make_shared<AudioParam>(
              options.Q,
              MOST_NEGATIVE_SINGLE_FLOAT,
              MOST_POSITIVE_SINGLE_FLOAT,
              context)),
      gainParam_(
          std::make_shared<AudioParam>(
              options.gain,
              MOST_NEGATIVE_SINGLE_FLOAT,
              BIQUAD_GAIN_DB_FACTOR * LOG10_MOST_POSITIVE_SINGLE_FLOAT,
              context)),
      computedFrequencyParam_(
          std::make_shared<CompositeAudioParam<combineBiquadFrequency>>(
              0.0f,
              getNyquistFrequency(),
              context,
              frequencyParam_,
              detuneParam_)),
      type_(options.type) {}

void BiquadFilterNode::setType(BiquadFilterType type) {
  type_ = type;
}

std::shared_ptr<AudioParam> BiquadFilterNode::getFrequencyParam() const {
  return frequencyParam_;
}

std::shared_ptr<AudioParam> BiquadFilterNode::getDetuneParam() const {
  return detuneParam_;
}

std::shared_ptr<AudioParam> BiquadFilterNode::getQParam() const {
  return QParam_;
}

std::shared_ptr<AudioParam> BiquadFilterNode::getGainParam() const {
  return gainParam_;
}

// Compute Z-transform of the filter
// https://www.dsprelated.com/freebooks/filters/Frequency_Response_Analysis.html
// https://www.dsprelated.com/freebooks/filters/Transfer_Function_Analysis.html
//
// frequency response -  H(z)
//          b0 + b1 * z^(-1) + b2 * z^(-2)
//  H(z) = -------------------------------
//           1 + a1 * z^(-1) + a2 * z^(-2)
//
//         b0 + (b1 + b2 * z1) * z1
//     =  --------------------------
//         (1 + (a1 + a2 * z1) * z1
//
// where z1 = 1/z and z = e^(j * pi * frequency)
// z1 = e^(-j * pi * frequency)
//
// phase response - angle of the frequency response
//

void BiquadFilterNode::getFrequencyResponse(
    const float *frequencyArray,
    float *magResponseOutput,
    float *phaseResponseOutput,
    const size_t length,
    BiquadFilterType type) {
  const float frequency = frequencyParam_->getValue();
  const double Q = QParam_->getValue();
  const double gain = gainParam_->getValue();
  const float detune = detuneParam_->getValue();

  const auto coeffs = applyFilter(combineBiquadFrequency(frequency, detune), Q, gain, type);

  const double nyquist = getNyquistFrequency();

  for (size_t i = 0; i < length; i++) {
    // Convert from frequency in Hz to normalized frequency [0, 1]
    const double normalizedFreq = static_cast<double>(frequencyArray[i]) / nyquist;

    if (normalizedFreq < 0.0 || normalizedFreq > 1.0) {
      // Out-of-bounds frequencies should return NaN.
      magResponseOutput[i] = std::nanf("");
      phaseResponseOutput[i] = std::nanf("");
      continue;
    }

    const double omega = kPi * normalizedFreq;
    const double cosW = std::cos(omega);
    const double sinW = std::sin(omega);
    const double cos2W = std::cos(2.0 * omega);
    const double sin2W = std::sin(2.0 * omega);

    // Match the decomposed H(e^{j*pi*f}) evaluation used by the WPT reference
    // (biquad-getFrequencyResponse.html / biquad-filters.js) for better
    // numerical agreement near Nyquist than complex division.
    const double numeratorReal = coeffs.b0 + coeffs.b1 * cosW + coeffs.b2 * cos2W;
    const double numeratorImag = -(coeffs.b1 * sinW + coeffs.b2 * sin2W);
    const double denominatorReal = 1.0 + coeffs.a1 * cosW + coeffs.a2 * cos2W;
    const double denominatorImag = -(coeffs.a1 * sinW + coeffs.a2 * sin2W);

    const double magnitude = std::sqrt(
        (numeratorReal * numeratorReal + numeratorImag * numeratorImag) /
        (denominatorReal * denominatorReal + denominatorImag * denominatorImag));
    magResponseOutput[i] = static_cast<float>(magnitude);

    double phase =
        std::atan2(numeratorImag, numeratorReal) - std::atan2(denominatorImag, denominatorReal);
    if (phase >= kPi) {
      phase -= 2.0 * kPi;
    } else if (phase <= -kPi) {
      phase += 2.0 * kPi;
    }
    phaseResponseOutput[i] = static_cast<float>(phase);
  }
}

BiquadFilterNode::FilterCoefficients BiquadFilterNode::getLowpassCoefficients(
    double frequency,
    double Q) {
  // Limit frequency to [0, 1] range
  if (frequency >= 1.0) {
    return getNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  if (frequency <= 0.0) {
    return getNormalizedCoefficients(0.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  const double g = std::pow(10.0, 0.05 * Q);

  const double theta = kPi * frequency;
  const double alpha = std::sin(theta) / (2 * g);
  const double cosW = std::cos(theta);
  const double beta = (1 - cosW) / 2;

  return getNormalizedCoefficients(beta, 2 * beta, beta, 1 + alpha, -2 * cosW, 1 - alpha);
}

BiquadFilterNode::FilterCoefficients BiquadFilterNode::getHighpassCoefficients(
    double frequency,
    double Q) {
  if (frequency >= 1.0) {
    return getNormalizedCoefficients(0.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }
  if (frequency <= 0.0) {
    return getNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  const double g = std::pow(10.0, 0.05 * Q);

  const double theta = kPi * frequency;
  const double alpha = std::sin(theta) / (2 * g);
  const double cosW = std::cos(theta);
  const double beta = (1 + cosW) / 2;

  return getNormalizedCoefficients(beta, -2 * beta, beta, 1 + alpha, -2 * cosW, 1 - alpha);
}

BiquadFilterNode::FilterCoefficients BiquadFilterNode::getBandpassCoefficients(
    double frequency,
    double Q) {
  // Limit frequency to [0, 1] range
  if (frequency <= 0.0 || frequency >= 1.0) {
    return getNormalizedCoefficients(0.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  // Limit Q to positive values
  if (Q <= 0.0) {
    return getNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  const double w0 = kPi * frequency;
  const double alpha = std::sin(w0) / (2 * Q);
  const double cosW = std::cos(w0);

  return getNormalizedCoefficients(alpha, 0.0, -alpha, 1.0 + alpha, -2 * cosW, 1.0 - alpha);
}

BiquadFilterNode::FilterCoefficients BiquadFilterNode::getLowshelfCoefficients(
    double frequency,
    double gain) {
  const double A = std::pow(10.0, gain / 40.0);

  if (frequency >= 1.0) {
    return getNormalizedCoefficients(A * A, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  if (frequency <= 0.0) {
    return getNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  const double w0 = kPi * frequency;
  const double alpha = 0.5 * std::sin(w0) * kSqrt2;
  const double cosW = std::cos(w0);
  const double gamma = 2.0 * std::sqrt(A) * alpha;

  return getNormalizedCoefficients(
      A * (A + 1 - (A - 1) * cosW + gamma),
      2.0 * A * (A - 1 - (A + 1) * cosW),
      A * (A + 1 - (A - 1) * cosW - gamma),
      A + 1 + (A - 1) * cosW + gamma,
      -2.0 * (A - 1 + (A + 1) * cosW),
      A + 1 + (A - 1) * cosW - gamma);
}

BiquadFilterNode::FilterCoefficients BiquadFilterNode::getHighshelfCoefficients(
    double frequency,
    double gain) {
  const double A = std::pow(10.0, gain / 40.0);

  if (frequency >= 1.0) {
    return getNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  if (frequency <= 0.0) {
    return getNormalizedCoefficients(A * A, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  const double w0 = kPi * frequency;
  // In the original formula: sqrt((A + 1/A) * (1/S - 1) + 2), but we assume
  // the maximum value S = 1, so it becomes 0 + 2 under the square root
  const double alpha = 0.5 * std::sin(w0) * kSqrt2;
  const double cosW = std::cos(w0);
  const double gamma = 2.0 * std::sqrt(A) * alpha;

  return getNormalizedCoefficients(
      A * (A + 1 + (A - 1) * cosW + gamma),
      -2.0 * A * (A - 1 + (A + 1) * cosW),
      A * (A + 1 + (A - 1) * cosW - gamma),
      A + 1 - (A - 1) * cosW + gamma,
      2.0 * (A - 1 - (A + 1) * cosW),
      A + 1 - (A - 1) * cosW - gamma);
}

BiquadFilterNode::FilterCoefficients
BiquadFilterNode::getPeakingCoefficients(double frequency, double Q, double gain) {
  const double A = std::pow(10.0, gain / 40.0);

  if (frequency <= 0.0 || frequency >= 1.0) {
    return getNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  if (Q <= 0.0) {
    return getNormalizedCoefficients(A * A, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  const double w0 = kPi * frequency;
  const double alpha = std::sin(w0) / (2 * Q);
  const double cosW = std::cos(w0);

  return getNormalizedCoefficients(
      1 + alpha * A, -2 * cosW, 1 - alpha * A, 1 + alpha / A, -2 * cosW, 1 - alpha / A);
}

BiquadFilterNode::FilterCoefficients BiquadFilterNode::getNotchCoefficients(
    double frequency,
    double Q) {
  if (frequency <= 0.0 || frequency >= 1.0) {
    return getNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  if (Q <= 0.0) {
    return getNormalizedCoefficients(0.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  const double w0 = kPi * frequency;
  const double alpha = std::sin(w0) / (2 * Q);
  const double cosW = std::cos(w0);

  return getNormalizedCoefficients(1.0, -2 * cosW, 1.0, 1 + alpha, -2 * cosW, 1 - alpha);
}

BiquadFilterNode::FilterCoefficients BiquadFilterNode::getAllpassCoefficients(
    double frequency,
    double Q) {
  if (frequency <= 0.0 || frequency >= 1.0) {
    return getNormalizedCoefficients(1.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  if (Q <= 0.0) {
    return getNormalizedCoefficients(-1.0, 0.0, 0.0, 1.0, 0.0, 0.0);
  }

  const double w0 = kPi * frequency;
  const double alpha = std::sin(w0) / (2 * Q);
  const double cosW = std::cos(w0);

  return getNormalizedCoefficients(
      1 - alpha, -2 * cosW, 1 + alpha, 1 + alpha, -2 * cosW, 1 - alpha);
}

BiquadFilterNode::FilterCoefficients BiquadFilterNode::getNormalizedCoefficients(
    double b0,
    double b1,
    double b2,
    double a0,
    double a1,
    double a2) {
  const double a0Inverted = 1.0 / a0;
  return {
      .b0 = b0 * a0Inverted,
      .b1 = b1 * a0Inverted,
      .b2 = b2 * a0Inverted,
      .a1 = a1 * a0Inverted,
      .a2 = a2 * a0Inverted};
}

BiquadFilterNode::FilterCoefficients BiquadFilterNode::applyFilter(
    double computedFrequency,
    double Q,
    double gain,
    BiquadFilterType type) {
  // `computedFrequency` already folds in detune (see `combineBiquadFrequency`).
  // NyquistFrequency is half of the sample rate.
  // Normalized frequency is therefore:
  // frequency / (sampleRate / 2) = (2 * frequency) / sampleRate
  double normalizedFrequency = computedFrequency / getNyquistFrequency();

  FilterCoefficients coeffs = {.b0 = 1.0, .b1 = 0.0, .b2 = 0.0, .a1 = 0.0, .a2 = 0.0};

  switch (type) {
    case BiquadFilterType::LOWPASS:
      coeffs = getLowpassCoefficients(normalizedFrequency, Q);
      break;
    case BiquadFilterType::HIGHPASS:
      coeffs = getHighpassCoefficients(normalizedFrequency, Q);
      break;
    case BiquadFilterType::BANDPASS:
      coeffs = getBandpassCoefficients(normalizedFrequency, Q);
      break;
    case BiquadFilterType::LOWSHELF:
      coeffs = getLowshelfCoefficients(normalizedFrequency, gain);
      break;
    case BiquadFilterType::HIGHSHELF:
      coeffs = getHighshelfCoefficients(normalizedFrequency, gain);
      break;
    case BiquadFilterType::PEAKING:
      coeffs = getPeakingCoefficients(normalizedFrequency, Q, gain);
      break;
    case BiquadFilterType::NOTCH:
      coeffs = getNotchCoefficients(normalizedFrequency, Q);
      break;
    case BiquadFilterType::ALLPASS:
      coeffs = getAllpassCoefficients(normalizedFrequency, Q);
      break;
    default:
      break;
  }

  return coeffs;
}

void BiquadFilterNode::processNode(int framesToProcess) {
  if (std::shared_ptr<BaseAudioContext> context = context_.lock()) {
    const auto currentTime = context->getCurrentTime();

    // k-rate: sample the parameters once per render quantum and reuse the
    // resulting coefficients for every frame in the block.
    const float computedFrequency = computedFrequencyParam_->processKRateParam(currentTime);
    const float Q = QParam_->processKRateParam(currentTime);
    const float gain = gainParam_->processKRateParam(currentTime);

    const auto coeffs = applyFilter(computedFrequency, Q, gain, type_);

    lastA2_ = coeffs.a2;

    const auto numChannels = audioBuffer_->getNumberOfChannels();

    for (size_t c = 0; c < numChannels; ++c) {
      // Filter state kept in double precision for the recursive difference
      // equation; only the sample I/O is float.
      double x1 = x1_[c];
      double x2 = x2_[c];
      double y1 = y1_[c];
      double y2 = y2_[c];

      auto channel = audioBuffer_->getChannel(c)->subSpan(framesToProcess);

      for (float &sample : channel) {
        const double input = sample;
        double output =
            coeffs.b0 * input + coeffs.b1 * x1 + coeffs.b2 * x2 - coeffs.a1 * y1 - coeffs.a2 * y2;

        // Avoid denormalized numbers
        if (std::abs(output) < 1e-15) {
          output = 0.0;
        }

        sample = static_cast<float>(output);

        x2 = x1;
        x1 = input;
        y2 = y1;
        y1 = output;
      }

      x1_[c] = x1;
      x2_[c] = x2;
      y1_[c] = y1;
      y2_[c] = y2;
    }
  } else {
    audioBuffer_->zero();
  }
}

int BiquadFilterNode::computeTailFrames() const {
  // some ai math slop, but it matches web audio api, so I think it's correct
  // Two safety bounds:
  //   1. `r` is clamped to `1 - kPoleRadiusEpsilon` so that `log(r)` stays
  //      bounded away from zero for poles sitting arbitrarily close to the
  //      unit circle (very high Q).
  //   2. The returned frame count is capped at `kMaxTailSeconds * sr` so a
  //      pathological coefficient update cannot keep a node alive forever.
  const double rRaw = std::sqrt(std::abs(lastA2_));
  const double r = std::min(rRaw, 1.0 - kPoleRadiusEpsilon);
  if (r <= 0.0) {
    return 0;
  }
  const double frames = std::ceil(std::log(kTailEpsilon) / std::log(r));
  if (!std::isfinite(frames) || frames <= 0.0) {
    return 0;
  }
  const int cap = static_cast<int>(kMaxTailSeconds * getContextSampleRate());
  return std::min(static_cast<int>(frames), cap);
}

} // namespace audioapi

// NOLINTEND(cppcoreguidelines-avoid-magic-numbers, readability-magic-numbers, readability-identifier-length)
